### DISCUSSION TITLE
Array spanning fields and registers

### DISCUSSION THREAD

**Discussion by maltaisn** [SystemRDL/systemrdl-compiler#204] [General]
I have an array of 8-bit values spanning 8 32-bit registers, for a total of 32 elements.

One way to describe it would be:
```systemrdl
  regfile {
    reg {
      field {} VALUE0[7:0];
      field {} VALUE1[15:8];
      field {} VALUE2[23:16];
      field {} VALUE3[31:24];
    } values[8];
  } my_array;
```
Another way:
```systemdl
  regfile {
    <% for ($i = 0; $i < 32; $i += 4) { %>
      reg {
        field {} VALUE<%=$i+0%>[7:0];
        field {} VALUE<%=$i+1%>[15:8];
        field {} VALUE<%=$i+2%>[23:16];
        field {} VALUE<%=$i+3%>[31:24];
      } values<%=$i%>;
    <% } %>
  } my_array;
```
In both cases the resulting C header, HTML documentation and hardware interface output aren't terribly interesting. Is there any other way of doing it? I couldn't find one. Ideally it would generate:
- C header: a `uint8_t arr[32]`
- HW interface: an array of 8-bit signals.

**Comment by amykyta3** [comment]
Regardless of how you implement it, the resulting register abstraction will still retain the structure of fields nested inside a register.
Your first implementation using an array of registers is definitely cleaner than using the Perl preprocessor.

Another option would be to use 8-bit registers, but that also implies your CPU bus interface is also narrower which may not be desirable:
```
  regfile {
    reg {
      regwidth = 8
      field {} VALUE[7:0];
    } values[32];
  } my_array;
```

**Reply by maltaisn** [reply]
Thank you.
